### ISSUE TITLE
Spanish Lib 0.3.3 Tasks List

### ISSUE THREAD

**Issue by tajmone** [alan-if/alan-i18n#46]
This Issue tracks the changes for the upcoming **Spanish Foundation v0.3.3**, built around PR #45 by @Rich15.

The task-list below will be updated and expanded as the work progresses. For features and topics being discussed elsewhere, reference links are provided to their respective _Discussion_ threads.

This Issue thread will host generic comments and discussion on the overall progress of the actual implementation work, but general considerations on the various features and changes, as affecting the overall **Foundation** project, or deciding what to include, should be dealt-with in dedicated _Discussion_ threads, to avoid cluttering this Issue, where we track decided-upon tasks.

- [x] `gramática.i`:
    + [x] _[Predefined Player Words]_ — Add missing SYNONYMS for:
        * [x] **AND WORDS** — `Y` and `LUEGO`.
        * [x] **BUT WORDS** — `excepto`
        * **THEM WORDS** — _postponed!_ (see #51)
    + [x] **Proper Named Entities** — entities with proper name have empty-string articles:
        * [x] Add new `nombre_propio` attribute (defaults to `Has not`).
        * [x] Tweak initialization code for entities with `nombre_propio` so they have empty-string articles (both `Definite`/`Indefinite`).
    + [x] Move plural pronouns from SYNONYMS of 'IT' to a separate definition matchin 'THEM' instead (see #51).
- [x] `mensajes.i` — polish and revise the translation of some Run-Time MESSAGEs (notes found in the source file):
    + [x] `CONTAINMENT_LOOP2`
    + [x] `NO_UNDO`
    + [x] `UNKNOWN_WORD`
- [x] **META VERBS**:
    + [x] `meta_sesión.i`:
        * [x] Add transcript verb?
        * **SYNONYMS for UNDO**? (see #47) — _not possible!_ it won't work, because UNDO is not part of the game dictionary, it's a special command intercepted by the parser.
- [x] `tests/meta.alan` — Implement all requied tests for META VERBs, Run-Time MESSAGEs and other "meta" stuff like _[predefined player words]_, etc.:
    + [x] @Rich15: Translate English & check Spanish in ALAN source.
    + [x] solution files:
        * [x] `meta-badcode.a3s`
        * [x] `meta-descriptions.a3s`
        * [x] `meta-input-errors.a3s`
        * [x] `meta-player-words.a3s`
        * [x] `meta-session.a3s`
    + [x] @tajmone: Examine carefully all `meta-*.a3t` transcripts for possible test improvements, formatting, etc..
        * [x] @Rich15: Once all the above meta tests are completed, could read all the transcripts again and check this box when no problems are found.
- [x] **PRE-MERGE TASKS** — Once all of the above are deal with:
    + [x] Update Lib Version & Date to **v0.3.3** (2021/11/22):
        * [x] `alan_es/README.md` &rarr; version badge.
        * [x] `alan_es/Foundation/CHANGELOG.md` &rarr; release entry.
        * [x] `alan_es/Foundation/Library.i` &rarr; header comments.
        * [x] `alan_es/vampiro/vampiro.alan` &rarr; header comments.
        * [x] Run Rake and commit changes.

-------------------------------------------------------------------------------

> **NOTE** — This next version might require more work than it might seem at first, because test coverage of meta verbs and run-time messages can be quite tricky to achieve, and as these tests are implemented they often reveal overlooked translation and GNA problems.
>
> Furthermore, some meta-feature inherited from **ALAN Lib v0.6** are worth discussing, since for the **Foundation Library** project (not just the Spanish library) we might improve or revise their design. Since we'll be looking closely into META VERBs, this might be a good time to evaluate and discuss them.

[Predefined Player Words]: https://github.com/alan-if/alan-i18n/wiki/Predefined-Player-Words

**Comment by tajmone** [comment]
# New Transcript META VERBs

@Rich15, I've added to `meta_sesión.i` new verbs to activate and stop saving a transcript of the game to file, so that players using a command line interpreter can access this feature:

```alan
Syntax script0    = transcripción.

Syntax script_on  = activar transcripción.
       script_on  = transcripción on.

Syntax script_off = desactivar transcripción.
       script_off = transcripción off.

Synonyms transcripcion, scripting, 'transcript' = transcripción.

Meta verb script0
  Does
    "Para activar la transcripción del juego, escriba: ACTIVAR TRANSCRIPCIÓN.
   $nPara terminar la transcripción, escriba: DESACTIVAR TRANSCRIPCIÓN."
End verb.

Meta verb script_on
  Does
    Transcript on. "La transcripción ya ha comenzado."
End verb.

Meta verb script_off
  Does
    Transcript off. "Fin de la transcripción."
End verb.
```

I did my best to come up with Spanish syntaxes and messages, but I'd rather have you check them and correct them if there are any problems with the text.

You can see the new verbs in action in the [`meta-session.a3t`][meta-session.a3t] test transcript.

I felt strongly about adding these verbs, otherwise command line players would have no means to activate the transcript feature.




[meta-session.a3t]: https://github.com/alan-if/alan-i18n/blob/dev_alan-es/alan_es/tests/meta-session.a3t#L154

**Comment by Rich15** [comment]
@tajmone, the verb works fine and all messages are well-written.

> I felt strongly about adding these verbs, otherwise command line players would have no means to activate the transcript feature.

Yes, it is a great addition!

**Comment by tajmone** [comment]
# Almost Ready to Merge …

@Rich15, I think we have enough material for a new Spanish release here.

Any further changes can go into the next release, unless you have some additional changes in mind which you thing should belong to this changeset.

## Translating New English Text in `meta.alan`

Since I had to update the `meta.alan` test adventure, adding a new actor, some objects and a verb in order to carry out some tests, there are some English sentences that we might want to translate, and my attempt to translate them might deserve double checking (all marked via `-- @TRANSLATE`).

## Adding UNDO Synonyms

I still have to add the **UNDO** SYNONYMS, just have to dig into the PR were we spoke about it. Also, if you have any further words to add to the SYNONYMS lists, just mention them here before we release.

## Releasing …

Then, once we agree that everything is ready, I just need to handle the final **PRE-MERGE TASKS** tasks, to update the library version info and date, the badges, etc., and we're good to squash merge the new version.

**Comment by Rich15** [comment]
> I still have to add the UNDO SYNONYMS, just have to dig into the PR were we spoke about it. Also, if you have any further words to add to the SYNONYMS lists, just mention them here before we release.

Right now I have no other Synonyms in mind besides the ones for 'IT' and 'THEM', but I think we already came to an agreement on those.

## Proper Named Entities

Before the next release I wanted to handle the named actors, so they don't appear with an article when listed.

What I'm planning to do is very similar to what you did in the Italian Library: defining a `nombre_propio` attribute and, if the entity has it, defining its article to none.

---

I'm doing that right now, so probably I will update the branch in a moment with the translations and the named actors handling.

**Comment by tajmone** [comment]
# UNDO Synonyms Won't Work

I discovered that it's not possible to define SYNONYMS for UNDO, because it wouldn't map to any author-defined words — i.e. the 'UNDO' word is not in the game dictionary, it's a special command intercepted by the parser.

So even with the SYNONYMS definition in place, in the actual game the parser would complaint that it doesn't know the word 'revierte', since it's not in the dictionary (or if it's there, it's mapping to an empty string).

I guess that UNDO belongs to that small group of hard-coded commands (along with RESTART and QUIT) which are not exposed to the author. But that's fine, after all they are well established and universally accepted terms among IF players, so there's no real need to translate them.

We tried...

# Proper Named Entities

Excellent job! Yes, it was a good idea to include also that feature in the next release.

**Comment by tajmone** [comment]
# PR Ready to Merge

@Rich15, I've prepared the final release as PR #52 and appointed you as reviewer. I've added some brief notes to. Feel free to rebase-merge when you're satisfied, or review the code.
